runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -9647,18 +9647,21 @@ J9::Z::TreeEvaluator::genGuardedLoadOOL(TR::Node *node, TR::CodeGenerator *cg, T
    }
 
 void
-J9::Z::TreeEvaluator::genArrayCopyWithArrayStoreCHK(TR::Node* node, TR::Register *srcObjReg, TR::Register *dstObjReg, TR::Register *srcAddrReg, TR::Register *dstAddrReg, TR::Register *lengthReg, TR::CodeGenerator *cg)
+J9::Z::TreeEvaluator::genArrayCopyWithArrayStoreCHK(TR::Node* node,
+                                                    TR::Register *srcObjReg,
+                                                    TR::Register *dstObjReg,
+                                                    TR::Register *srcAddrReg,
+                                                    TR::Register *dstAddrReg,
+                                                    TR::Register *lengthReg,
+                                                    TR::CodeGenerator *cg,
+                                                    bool doResultCheck)
```

Comment:
I'm adding `bool doResultCheck` to this API. The rest of changes in this function are just code motion.
`bool doResultCheck` is needed because some arrayCopy nodes do not need store checks as `node->isNoArrayStoreCheckArrayCopy()` query is true. Hence, we need to optionally generate the store check sequence, which raises exceptions.

Reply:
I don't see a reason why we should modify this API. The intent is clear. `genArrayCopyWithArrayStoreCHK` will generate an array copy sequence _with_ an array store check at the end. If you don't need an array store check you should be calling the primitive array copy evaluator.

Looking at the locations this API is called what I think you really wanted is to "wrap" the call to `OMR::TreeEvaluatorConnector::arraycopyEvaluator`. i.e. something like this:

```
TR::Register *
J9::Z::TreeEvaluator::referenceArraycopyEvaluator(TR::Node* node, TR::CodeGenerator* cg, TR::Node* byteSrcNode, TR::Node* byteDstNode, TR::Node* byteLenNode, TR::Node* byteSrcObjNode, TR::Node* byteDstObjNode)
   {
   if (TR::Compiler->om.shouldReplaceGuardedLoadWithSoftwareReadBarrier())
      {
      TR::Node* byteSrcObjNode = node->getChild(0);
      TR::Node* byteDstObjNode = node->getChild(1);

      TR::Node* byteSrcNode = node->getChild(2);
      TR::Node* byteDstNode = node->getChild(3);
      TR::Node* byteLenNode = node->getChild(4);

      // Check if we need to call VM helper
      // Create VM helper call in OOL, etc.

      auto result = OMR::TreeEvaluatorConnector::referenceArraycopyEvaluator(node, cg, byteSrcNode, byteDstNode, byteLenNode, byteSrcObjNode, byteDstObjNode);

      // Generate done label

      return result;
      }
   else
      {
      return OMR::TreeEvaluatorConnector::referenceArraycopyEvaluator(node, cg, byteSrcNode, byteDstNode, byteLenNode, byteSrcObjNode, byteDstObjNode);
      }
   }
```

We should not be overriding `TreeEvaluator::arraycopyEvaluator` IMO as this is not what we really want. In general try not to modify existing APIs if possible.

Reply:
This is addressed in 85425a34db.

Reply:
I don't think this is addressed. I do not believe we should be complicating this API with the `doResultCheck` parameter. This API will generate an array copy sequence _with_ an array store check at the end. If you don't need an array store check you should be calling the primitive array copy evaluator. See above example for how the `referenceArraycopyEvaluator` function can be implemented without having to change the `genArrayCopyWithArrayStoreCHK` API at all.

Reply:
> If you don't need an array store check you should be calling the primitive array copy evaluator.

But we plan to replace load-store loops with helper calls in concurrent scavenge mode, don't we?

`genArrayCopyWithArrayStoreCHK` and `primitiveArrayCopy` generate two different sequences. The former constructs a helper call with store check (after call returns) and the latter does load-store loops.

These two are not really interchangable in my situation. For SW-CS,  I need a helper call when it's in GC cycle; for HW-CS, I'm replacing load-store-loop with helper call and planning to measure their performance difference.


I can get away without modifying `genArrayCopyWithArrayStoreCHK` API (no extra bool param), but the body of `genArrayCopyWithArrayStoreCHK` needs a small change:  generate store check instructions only when `!node->chkNoArrayStoreCheckArrayCopy()` is true. After all, the store check is needed only when this flag is true.


As for `referenceArraycopyEvaluator`(in OMR and J9), given that OMR has no notion of reference and should not have reference array handling. I'll absort the body of `OMR::TreeEvaluatorConnector::referenceArraycopyEvaluator`
into `J9::Z::TreeEvaluator::referenceArraycopyEvaluator`. OMR `referenceArraycopyEvaluator()` will be empty and be kept there for other projects to override.
`OMR::TreeEvaluatorConnector::arrayCopyEvaluator()` will call `referenceArraycopyEvaluator` or `primitiveArrayCopyEvaluator` depending on the arrayCopy node type.


`J9::Z::referenceArraycopyEvaluator` needs to handle four cases:
1. hardware CS: this was previously implemented as a load-store loop and has been changed to a helper call. Perf evaluation is to be carried out.
2. software CS and no need for store check: helper call in the OOL and primitive array copy in mainline code.
3. array needs store check in any GC mode: call helper in mainline.
4. non-CS mode and no need for store check: primitive array copy with wrtbar.

Overall, the code I will push looks like this:

```
TR::Register *
J9::Z::TreeEvaluator::referenceArraycopyEvaluator(...)
   {
   
   // Call helper if:
   // 1. it's hardware concurrent scavenge mode
   // 2. it's a reference array copy that need an array store check
   //
   // If not in CS mode, use regular load-store loop.
   if (!TR::Compiler->om.shouldReplaceGuardedLoadWithSoftwareReadBarrier()
       || !node->chkNoArrayStoreCheckArrayCopy())
      {
	  if (!node->chkNoArrayStoreCheckArrayCopy() || TR::Compiler->om.shouldGenerateReadBarriersForFieldLoads()) 
	     {
		 // create VM helper call in mainline
		 }
	  else
	     {
	     // when the JIT is not in CS mode AND chkNoArrayStoreCheckArrayCopy() is true
		 // call primitive array copy (load-store loop) with wrtbar
	     }

      return result;
      }
   else
      {
	  // SW CS here
	  
	  // check if in CS cycle
	  // create VM helper call in OOL. executed when in CS cycle
	  // create primiteive array copy sequence with wrtbar in mainline. executed when not in CS cycle
	  
	  // done
	  
	  return result;
      }
   }
```

Reply:
Assuming the performance of calling the helper in the HW CS mode is good I don't understand why the HW and SW paths are different at all? Shouldn't they be the same? i.e. if you are in a CS cycle call the VM helper and we're done, otherwise fall through to execute the exact same arraycopy sequence had CS not even be present?

Can you please explain why we cannot have 2 cases:

1. CS is active
  - Goto OOL and call VM helper always irregardless of HW or SW CS
2. CS is not active
  - Fall through and carry out exact same sequence as if CS were not active at all

Reply:
cb0cc1101924819144a7bfb80ad491448e466abb and OMR change 6c68b40c7f273c11326a9eb44696dd1f29c39ae5 attempt to fix reference array copy in HW CS mode.

HW and SW CS reference array copy are implemented the same.

The above commits also remove `genGuardedLoadOOL` and some related code because OOL guarded-load-store loops are replaced by OOL helper calls. Performance evaluation is yet to be done.